src/views/ApartmentApp/components/StudentApplication/index.js

```diff
@@ -0,0 +1,794 @@
+  const [snackbarOpen, setSnackbarOpen] = useState(false);
+  const [snackbarText, setSnackbarText] = useState('');
+  const [snackbarSeverity, setSnackbarSeverity] = useState('info');
+  const [saveButtonAlertTimeout, setSaveButtonAlertTimeout] = useState(null);
+
+  /**
+   * Loads the user's saved apartment application, if one exists
+   */
+  useEffect(() => {
+    let isSubscribed = true;
+
+    const handleError = (_error) => {
+      // No existing application was found in the database,
+      // or an error occurred while attempting to load the application
+      setApplicationID(null);
+      setEditorUsername((prevEditorUsername) => prevEditorUsername ?? userProfile.AD_Username);
+      setApplicants((prevApplicants) =>
+        prevApplicants.length ? prevApplicants : [{ Profile: userProfile, OffCampusProgram: '' }],
+      );
+      setUnsavedChanges(true);
+    };
+
+    if (!applicationID || !applicationDetails) {
+      setLoading(true);
+      // Check if the current user is on an application. Returns the application ID number if found
+      housing
+        .getCurrentApplicationID()
+        .then((newApplicationID) => {
+          console.log('Retrieved Application ID: ' + newApplicationID); //! DEBUG
+          if (newApplicationID === null || newApplicationID === -1) {
+            // Intentionally trigger the 'catch'
+            throw new Error("Invalid value of 'newApplicationID' = " + newApplicationID);
+          } else if (isSubscribed) {
+            setApplicationID(newApplicationID);
+            setUnsavedChanges(false);
+            housing
+              .getApartmentApplication(newApplicationID)
+              .then((newApplicationDetails) => {
+                if (isSubscribed && newApplicationDetails) {
+                  setApplicationDetails(newApplicationDetails);
+                  setDateSubmitted(newApplicationDetails.DateSubmitted ?? null);
+                  setDateModified(newApplicationDetails.DateModified ?? null);
+                  setEditorUsername(newApplicationDetails.EditorUsername ?? null);
+                  setApplicants((prevApplicants) => {
+                    if (newApplicationDetails.Applicants?.length > 0) {
+                      return newApplicationDetails.Applicants.map(async (newApplicantInfo) => {
+                        if (newApplicantInfo.Profile !== null) {
+                          return newApplicantInfo;
+                        } else {
+                          const newApplicantProfile = await user.getProfileInfo(
+                            newApplicantInfo.Username,
+                          );
+                          return { Profile: newApplicantProfile, ...newApplicantInfo };
+                        }
+                      });
+                    } else {
+                      return prevApplicants;
+                    }
+                  });
```

Comment:
This does not work, because your arrow function supplied to map is asynchronous, you're actually returning a Promise. Also, since this is only running when the page first loads (or userProfile changes), you know that `applicants` will always be an empty array, so you don't need to check for the case that it isn't.
```suggestion
                  Promise.all(
                    newApplicationDetails.Applicants.map(async (newApplicantInfo) => {
                      if (newApplicantInfo.Profile !== null) {
                        return newApplicantInfo;
                      } else {
                        const newApplicantProfile = await user.getProfileInfo(
                          newApplicantInfo.Username,
                        );
                        return { Profile: newApplicantProfile, ...newApplicantInfo };
                      }
                    }),
                  ).then((newApplicants) => setApplicants(newApplicants));
```

Reply:
Also, is it possible for an `Applicant` returned from the database to have no `Profile` attached? If not, than all this is unnecessary, and if so than maybe you should fix that instead?